src/utils/__tests__/internalLinks-test.js

```diff
@@ -10,157 +11,84 @@ import {
-    [false, 'same domain, triple slash', 'https://example.com///#narrow/stream/jest'],
-
-    // These examples may be odd-looking URLs, but are nevertheless valid.
-    [true, 'scheme-relative', '//example.com/#narrow/stream/jest'],
-    [true, 'path-relative, nop path', '.#narrow/stream/jest'],
-    [true, 'path-relative, nop path', './#narrow/stream/jest'],
-    [true, 'path-relative, nop path', '..#narrow/stream/jest'],
-    [true, 'path-relative, nop path', '../#narrow/stream/jest'],
-    [true, 'path-relative, nop path', 'foo/..#narrow/stream/jest'],
-    [true, 'path-relative, nop path', 'foo/../#narrow/stream/jest'],
-    [true, 'path-relative, nop path', 'foo/bar/../../../baz/.././#narrow/stream/jest'],
-    [true, 'path-absolute, nop path', '/.#narrow/stream/jest'],
-    [true, 'path-absolute, nop path', '/./#narrow/stream/jest'],
-    [true, 'path-absolute, nop path', '/..#narrow/stream/jest'],
-    [true, 'path-absolute, nop path', '/../#narrow/stream/jest'],
-    [true, 'path-absolute, nop path', '/foo/..#narrow/stream/jest'],
-    [true, 'path-absolute, nop path', '/foo/../#narrow/stream/jest'],
-    [true, 'path-absolute, nop path', '/foo/bar/../../../baz/.././#narrow/stream/jest'],
-    [true, 'same domain, nop path', 'https://example.com/.#narrow/stream/jest'],
-    [true, 'same domain, nop path', 'https://example.com/./#narrow/stream/jest'],
-    [true, 'same domain, nop path', 'https://example.com/..#narrow/stream/jest'],
-    [true, 'same domain, nop path', 'https://example.com/../#narrow/stream/jest'],
-    [true, 'same domain, nop path', 'https://example.com/foo/..#narrow/stream/jest'],
-    [true, 'same domain, nop path', 'https://example.com/foo/../#narrow/stream/jest'],
-    [
-      true,
-      'same domain, nop path',
-      'https://example.com/foo/bar/../../../baz/.././#narrow/stream/jest',
-    ],
-    [true, 'same domain, %-encoded host', 'https://%65xample%2ecom/#narrow/stream/jest'],
-    // This one fails because our polyfilled URL implementation has IDNA stripped out.
-    // [true, 'same domain, punycoded host', 'https://example.xn--h2brj9c/#narrow/stream/jest', new URL('https://example.भारत/'),], // FAILS
-    [
-      true,
-      'same domain, punycodable host',
-      'https://example.भारत/#narrow/stream/jest',
-      new URL('https://example.भारत/'),
-    ],
-    // This one fails because our polyfilled URL implementation has IDNA stripped out.
-    // [true, 'same domain, IDNA-mappable', 'https://ℯⅩªm🄿ₗℰ.ℭᴼⓂ/#narrow/stream/jest'], // FAILS
-    [
-      true,
-      'same IPv4 address, %-encoded',
-      'http://%31%39%32%2e168%2e0%2e1/#narrow/stream/jest',
-      new URL('http://192.168.0.1/'),
-    ],
-    // This one fails because our polyfilled URL implementation has IDNA stripped out.
-    // [true, 'same IPv4 address, IDNA-mappable', 'http://１𝟗𝟚。①⁶🯸．₀｡𝟭/#narrow/stream/jest', new URL('http://192.168.0.1/'),], // FAILS
```

Comment:
OK, I think all three of these are still as relevant as before and should be retained.

They're really about limitations in our URL implementation, not particularly specific to `isNarrowLink`. But this is the spot where we have the most systematic URL-related tests, so it's as good a place as any for them to live. And they do affect the overall behavior we get when we invoke `isNarrowLink` -- it's just that things will have already gone wrong before we call the latter, when we invoked `new URL`.

What each of these is describing is a situation where if we're given a URL string of a certain form -- perhaps in a Zulip message somewhere -- we can look at it and come to one conclusion about what it means, and a different URL implementation like the one in the user's browser would come to a different conclusion.

That's always a concerning situation in general, because it's the sort of thing that can lead to a class of security vulnerabilities: we look at the URL and decide it's OK to do thing X with it because it meets constraint Y, hand it off to a browser or something to carry out thing X, and by the browser's lights it doesn't actually meet constraint Y so it ends up doing some totally other thing ZZZ that we didn't intend.

(I'm not aware of any concrete way this particular issue gets any closer than that to causing a live security vulnerability. If you do see such a way, please [report it confidentially](https://github.com/zulip/zulip/blob/main/SECURITY.md); the above is already right on the edge of getting uncomfortable to discuss in public.)

Specifically, with one of these URL strings like `http://１𝟗𝟚。①⁶🯸．₀｡𝟭/`, a correct URL implementation will produce the same result as for `http://192.168.0.1/`. The one we're using will miss that transformation, and will think the origin is `http://１𝟗𝟚。①⁶🯸．₀｡𝟭`.